### PR TITLE
Use vec for storing edges

### LINKED ISSUE #19: Optimize Adjacency Tracking
- Resizing of HashSet while iterating is very expensive

### PR DESCRIPTION
**Pull request by apfitzge**
## Problem 
#19 - Adjacency tracking performance is poor using a HashSet; resizing, hashing, etc.

## Changes

- Use a Vec to store edges instead of HashSet.
- Check only last item in vec, since we check only check edge uniqueness on insertion
- Fixes short-coming of #79. The public API now uses `Vec<Id>` as return type; no reliance on `ahash` types for exposed API.

Fixes #19

### CODE REVIEW

**Comment by apfitzge** [comment]
A few cases with minor performance regression, within noise threshold.
Otherwise seeing reasonably good performance gains, particularly when there is a lot of conflicts.

<details>
<summary>Benchmark Results</summary>

```bash
no_conflict_100_2       time:   [7.5968 µs 7.6007 µs 7.6050 µs]
                        change: [-12.436% -12.144% -11.824%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 13 outliers among 100 measurements (13.00%)
  2 (2.00%) high mild
  11 (11.00%) high severe

no_conflict_100_16      time:   [34.753 µs 35.382 µs 36.022 µs]
                        change: [-4.1051% -2.6098% -1.0759%] (p = 0.00 < 0.05)
                        Performance has improved.

no_conflict_100_32      time:   [66.204 µs 67.330 µs 68.256 µs]
                        change: [-10.036% -8.5761% -6.9666%] (p = 0.00 < 0.05)
                        Performance has improved.

no_conflict_100_64      time:   [134.99 µs 137.48 µs 139.64 µs]
                        change: [-6.6239% -4.6241% -2.5877%] (p = 0.00 < 0.05)
                        Performance has improved.

no_conflict_100_128     time:   [309.01 µs 310.24 µs 311.61 µs]
                        change: [-1.7082% -0.9979% -0.2679%] (p = 0.01 < 0.05)
                        Change within noise threshold.
Found 7 outliers among 100 measurements (7.00%)
  5 (5.00%) high mild
  2 (2.00%) high severe

no_conflict_100_256     time:   [629.21 µs 635.35 µs 642.27 µs]
                        change: [+0.7081% +1.7161% +2.7499%] (p = 0.00 < 0.05)
                        Change within noise threshold.

no_conflict_1000_2      time:   [106.30 µs 106.90 µs 107.54 µs]
                        change: [-11.579% -10.977% -10.396%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 11 outliers among 100 measurements (11.00%)
  10 (10.00%) high mild
  1 (1.00%) high severe

no_conflict_1000_16     time:   [594.77 µs 602.28 µs 610.53 µs]
                        change: [+0.3372% +1.3480% +2.3462%] (p = 0.01 < 0.05)
                        Change within noise threshold.
Found 7 outliers among 100 measurements (7.00%)
  3 (3.00%) high mild
  4 (4.00%) high severe

no_conflict_1000_32     time:   [1.1254 ms 1.1334 ms 1.1420 ms]
                        change: [-0.2618% +0.9449% +2.1731%] (p = 0.14 > 0.05)
                        No change in performance detected.
Found 13 outliers among 100 measurements (13.00%)
  6 (6.00%) high mild
  7 (7.00%) high severe

no_conflict_1000_64     time:   [1.9341 ms 1.9402 ms 1.9460 ms]
                        change: [-3.0565% -2.3975% -1.5266%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 5 outliers among 100 measurements (5.00%)
  2 (2.00%) high mild
  3 (3.00%) high severe

no_conflict_1000_128    time:   [3.7179 ms 3.7285 ms 3.7393 ms]
                        change: [-2.8074% -2.3022% -1.8249%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 1 outliers among 100 measurements (1.00%)
  1 (1.00%) high mild

no_conflict_1000_256    time:   [9.4015 ms 9.4481 ms 9.5042 ms]
                        change: [-4.6442% -4.0796% -3.4998%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 10 outliers among 100 measurements (10.00%)
  10 (10.00%) high severe

no_conflict_10000_2     time:   [1.1152 ms 1.1322 ms 1.1482 ms]
                        change: [-5.6260% -3.5611% -1.5744%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 30 outliers among 100 measurements (30.00%)
  20 (20.00%) low severe
  5 (5.00%) high mild
  5 (5.00%) high severe

no_conflict_10000_16    time:   [4.9481 ms 4.9662 ms 4.9841 ms]
                        change: [-0.4987% +0.5761% +1.6435%] (p = 0.30 > 0.05)
                        No change in performance detected.
Found 8 outliers among 100 measurements (8.00%)
  1 (1.00%) low severe
  1 (1.00%) low mild
  4 (4.00%) high mild
  2 (2.00%) high severe

no_conflict_10000_32    time:   [11.395 ms 11.432 ms 11.476 ms]
                        change: [-0.5371% +0.1683% +0.8306%] (p = 0.64 > 0.05)
                        No change in performance detected.
Found 8 outliers among 100 measurements (8.00%)
  4 (4.00%) high mild
  4 (4.00%) high severe

no_conflict_10000_64    time:   [26.077 ms 26.160 ms 26.252 ms]
                        change: [-13.207% -11.999% -10.798%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 7 outliers among 100 measurements (7.00%)
  5 (5.00%) high mild
  2 (2.00%) high severe

no_conflict_10000_128   time:   [58.500 ms 58.676 ms 58.868 ms]
                        change: [-2.6857% -2.2598% -1.8273%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 6 outliers among 100 measurements (6.00%)
  5 (5.00%) high mild
  1 (1.00%) high severe

no_conflict_10000_256   time:   [128.33 ms 129.50 ms 130.80 ms]
                        change: [+0.5411% +1.5834% +2.7285%] (p = 0.00 < 0.05)
                        Change within noise threshold.
Found 11 outliers among 100 measurements (11.00%)
  5 (5.00%) high mild
  6 (6.00%) high severe

all_conflict_100_2      time:   [15.254 µs 15.348 µs 15.443 µs]
                        change: [-23.695% -23.436% -23.144%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 19 outliers among 100 measurements (19.00%)
  1 (1.00%) low mild
  6 (6.00%) high mild
  12 (12.00%) high severe

all_conflict_100_16     time:   [48.864 µs 49.023 µs 49.222 µs]
                        change: [-19.958% -19.682% -19.412%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 11 outliers among 100 measurements (11.00%)
  8 (8.00%) high mild
  3 (3.00%) high severe

all_conflict_100_32     time:   [87.238 µs 87.329 µs 87.441 µs]
                        change: [-19.951% -19.755% -19.559%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 8 outliers among 100 measurements (8.00%)
  5 (5.00%) high mild
  3 (3.00%) high severe

all_conflict_100_64     time:   [163.62 µs 163.93 µs 164.28 µs]
                        change: [-20.018% -19.813% -19.628%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 13 outliers among 100 measurements (13.00%)
  4 (4.00%) high mild
  9 (9.00%) high severe

all_conflict_100_128    time:   [317.57 µs 317.97 µs 318.41 µs]
                        change: [-19.097% -18.904% -18.722%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 8 outliers among 100 measurements (8.00%)
  4 (4.00%) high mild
  4 (4.00%) high severe

all_conflict_100_256    time:   [616.99 µs 618.10 µs 619.57 µs]
                        change: [-18.893% -18.652% -18.412%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 4 outliers among 100 measurements (4.00%)
  1 (1.00%) high mild
  3 (3.00%) high severe

all_conflict_1000_2     time:   [146.57 µs 146.80 µs 147.08 µs]
                        change: [-31.427% -31.054% -30.653%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 1 outliers among 100 measurements (1.00%)
  1 (1.00%) high mild

all_conflict_1000_16    time:   [477.70 µs 477.95 µs 478.22 µs]
                        change: [-23.551% -23.273% -22.970%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 8 outliers among 100 measurements (8.00%)
  3 (3.00%) high mild
  5 (5.00%) high severe

all_conflict_1000_32    time:   [855.47 µs 856.76 µs 858.33 µs]
                        change: [-22.638% -22.388% -22.151%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 2 outliers among 100 measurements (2.00%)
  2 (2.00%) high mild

all_conflict_1000_64    time:   [1.6124 ms 1.6147 ms 1.6171 ms]
                        change: [-21.905% -21.682% -21.482%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 14 outliers among 100 measurements (14.00%)
  5 (5.00%) high mild
  9 (9.00%) high severe

all_conflict_1000_128   time:   [3.1459 ms 3.1512 ms 3.1568 ms]
                        change: [-20.971% -20.788% -20.611%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 8 outliers among 100 measurements (8.00%)
  8 (8.00%) high mild

all_conflict_1000_256   time:   [6.1092 ms 6.1208 ms 6.1338 ms]
                        change: [-20.642% -20.448% -20.250%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 5 outliers among 100 measurements (5.00%)
  3 (3.00%) high mild
  2 (2.00%) high severe

all_conflict_10000_2    time:   [1.5245 ms 1.5322 ms 1.5391 ms]
                        change: [-28.645% -28.013% -27.395%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 24 outliers among 100 measurements (24.00%)
  16 (16.00%) low severe
  2 (2.00%) low mild
  2 (2.00%) high mild
  4 (4.00%) high severe

all_conflict_10000_16   time:   [4.8649 ms 4.8770 ms 4.8889 ms]
                        change: [-23.410% -23.019% -22.630%] (p = 0.00 < 0.05)
                        Performance has improved.

all_conflict_10000_32   time:   [8.6764 ms 8.6889 ms 8.7023 ms]
                        change: [-21.978% -21.623% -21.284%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 6 outliers among 100 measurements (6.00%)
  2 (2.00%) low mild
  2 (2.00%) high mild
  2 (2.00%) high severe

all_conflict_10000_64   time:   [16.136 ms 16.161 ms 16.188 ms]
                        change: [-21.907% -21.655% -21.415%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 2 outliers among 100 measurements (2.00%)
  2 (2.00%) high mild

all_conflict_10000_128  time:   [31.756 ms 31.805 ms 31.861 ms]
                        change: [-20.093% -19.908% -19.701%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 5 outliers among 100 measurements (5.00%)
  1 (1.00%) low mild
  2 (2.00%) high mild
  2 (2.00%) high severe

all_conflict_10000_256  time:   [61.511 ms 61.618 ms 61.725 ms]
                        change: [-20.097% -19.892% -19.698%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 1 outliers among 100 measurements (1.00%)
  1 (1.00%) high mild

tree_100_4_16           time:   [50.050 µs 50.091 µs 50.140 µs]
                        change: [-17.067% -16.862% -16.655%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 9 outliers among 100 measurements (9.00%)
  2 (2.00%) high mild
  7 (7.00%) high severe

tree_100_8_16           time:   [50.744 µs 50.826 µs 50.910 µs]
                        change: [-16.369% -16.116% -15.879%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 9 outliers among 100 measurements (9.00%)
  3 (3.00%) high mild
  6 (6.00%) high severe

tree_100_4_64           time:   [164.00 µs 164.26 µs 164.53 µs]
                        change: [-18.990% -18.828% -18.660%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 18 outliers among 100 measurements (18.00%)
  1 (1.00%) low mild
  4 (4.00%) high mild
  13 (13.00%) high severe

tree_100_8_64           time:   [173.24 µs 175.44 µs 177.39 µs]
                        change: [-16.636% -16.082% -15.506%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 21 outliers among 100 measurements (21.00%)
  1 (1.00%) low mild
  1 (1.00%) high mild
  19 (19.00%) high severe

tree_100_4_128          time:   [319.37 µs 319.85 µs 320.36 µs]
                        change: [-18.152% -17.990% -17.830%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 17 outliers among 100 measurements (17.00%)
  7 (7.00%) high mild
  10 (10.00%) high severe

tree_100_8_128          time:   [336.30 µs 338.23 µs 339.93 µs]
                        change: [-17.211% -16.628% -15.995%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 2 outliers among 100 measurements (2.00%)
  2 (2.00%) high mild

tree_1000_4_16          time:   [504.12 µs 504.30 µs 504.53 µs]
                        change: [-20.254% -19.946% -19.653%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 10 outliers among 100 measurements (10.00%)
  2 (2.00%) high mild
  8 (8.00%) high severe

tree_1000_8_16          time:   [504.39 µs 505.69 µs 507.32 µs]
                        change: [-18.989% -18.633% -18.303%] (p = 0.00 < 0.05)
                        Performance has improved.

tree_1000_4_64          time:   [1.6691 ms 1.6718 ms 1.6750 ms]
                        change: [-19.042% -18.818% -18.573%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 8 outliers among 100 measurements (8.00%)
  2 (2.00%) low mild
  6 (6.00%) high severe

tree_1000_8_64          time:   [1.6712 ms 1.6731 ms 1.6752 ms]
                        change: [-19.522% -19.323% -19.118%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 4 outliers among 100 measurements (4.00%)
  2 (2.00%) high mild
  2 (2.00%) high severe

tree_1000_4_128         time:   [3.2514 ms 3.2578 ms 3.2653 ms]
                        change: [-19.058% -18.784% -18.532%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 13 outliers among 100 measurements (13.00%)
  3 (3.00%) high mild
  10 (10.00%) high severe

tree_1000_8_128         time:   [3.2216 ms 3.2241 ms 3.2269 ms]
                        change: [-19.382% -19.193% -19.021%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 3 outliers among 100 measurements (3.00%)
  3 (3.00%) high severe

random_access_100_2_2   time:   [15.937 µs 16.103 µs 16.246 µs]
                        change: [-22.921% -22.194% -21.487%] (p = 0.00 < 0.05)
                        Performance has improved.

random_access_100_100_2 time:   [14.214 µs 14.237 µs 14.262 µs]
                        change: [-19.247% -18.975% -18.696%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 19 outliers among 100 measurements (19.00%)
  3 (3.00%) low mild
  8 (8.00%) high mild
  8 (8.00%) high severe

random_access_100_100_16
                        time:   [90.241 µs 90.546 µs 90.855 µs]
                        change: [-17.335% -16.553% -15.920%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 2 outliers among 100 measurements (2.00%)
  1 (1.00%) high mild
  1 (1.00%) high severe

random_access_100_100_32
                        time:   [130.29 µs 130.44 µs 130.59 µs]
                        change: [-17.851% -17.615% -17.381%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 13 outliers among 100 measurements (13.00%)
  4 (4.00%) low mild
  4 (4.00%) high mild
  5 (5.00%) high severe

random_access_100_100_64
                        time:   [197.93 µs 198.18 µs 198.56 µs]
                        change: [-17.856% -17.642% -17.406%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 13 outliers among 100 measurements (13.00%)
  1 (1.00%) low mild
  4 (4.00%) high mild
  8 (8.00%) high severe

random_access_100_1000_2
                        time:   [9.4872 µs 9.4962 µs 9.5067 µs]
                        change: [-13.330% -13.084% -12.829%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 7 outliers among 100 measurements (7.00%)
  5 (5.00%) high mild
  2 (2.00%) high severe

random_access_100_1000_16
                        time:   [67.344 µs 67.444 µs 67.528 µs]
                        change: [-22.851% -21.707% -20.590%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 6 outliers among 100 measurements (6.00%)
  1 (1.00%) low mild
  4 (4.00%) high mild
  1 (1.00%) high severe

random_access_100_1000_32
                        time:   [149.70 µs 149.77 µs 149.87 µs]
                        change: [-21.313% -20.431% -19.495%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 13 outliers among 100 measurements (13.00%)
  3 (3.00%) high mild
  10 (10.00%) high severe

random_access_100_1000_64
                        time:   [284.14 µs 284.28 µs 284.43 µs]
                        change: [-20.686% -20.148% -19.500%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 16 outliers among 100 measurements (16.00%)
  4 (4.00%) high mild
  12 (12.00%) high severe

random_access_100_1000_128
                        time:   [455.72 µs 457.38 µs 459.38 µs]
                        change: [-18.460% -18.102% -17.743%] (p = 0.00 < 0.05)
                        Performance has improved.

random_access_100_1000_256
                        time:   [799.53 µs 804.11 µs 807.99 µs]
                        change: [-17.002% -16.659% -16.234%] (p = 0.00 < 0.05)
                        Performance has improved.

random_access_100_10000_2
                        time:   [7.9586 µs 7.9643 µs 7.9696 µs]
                        change: [-18.278% -18.052% -17.836%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 10 outliers among 100 measurements (10.00%)
  2 (2.00%) low mild
  5 (5.00%) high mild
  3 (3.00%) high severe

random_access_100_10000_16
                        time:   [46.324 µs 48.223 µs 50.129 µs]
                        change: [-10.024% -6.2804% -2.4907%] (p = 0.00 < 0.05)
                        Performance has improved.

random_access_100_10000_32
                        time:   [109.45 µs 110.09 µs 110.83 µs]
                        change: [-5.3487% -3.8711% -2.3261%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 1 outliers among 100 measurements (1.00%)
  1 (1.00%) high mild

random_access_100_10000_64
                        time:   [242.42 µs 243.01 µs 243.68 µs]
                        change: [-6.4505% -5.6100% -4.7099%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 7 outliers among 100 measurements (7.00%)
  4 (4.00%) high mild
  3 (3.00%) high severe

random_access_100_10000_128
                        time:   [605.05 µs 615.52 µs 624.82 µs]
                        change: [-3.4741% -1.9877% -0.3940%] (p = 0.02 < 0.05)
                        Change within noise threshold.
Found 18 outliers among 100 measurements (18.00%)
  14 (14.00%) low severe
  1 (1.00%) low mild
  2 (2.00%) high mild
  1 (1.00%) high severe

random_access_100_10000_256
                        time:   [1.0380 ms 1.0412 ms 1.0436 ms]
                        change: [-15.214% -14.240% -13.257%] (p = 0.00 < 0.05)
                        Performance has improved.

random_access_1000_2_2  time:   [147.66 µs 148.54 µs 149.54 µs]
                        change: [-30.215% -29.619% -29.022%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 22 outliers among 100 measurements (22.00%)
  13 (13.00%) low mild
  3 (3.00%) high mild
  6 (6.00%) high severe

random_access_1000_100_2
                        time:   [178.20 µs 178.38 µs 178.58 µs]
                        change: [-24.097% -23.560% -23.013%] (p = 0.00 < 0.05)
                        Performance has improved.

random_access_1000_100_16
                        time:   [967.24 µs 969.76 µs 972.61 µs]
                        change: [-15.903% -15.594% -15.223%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 27 outliers among 100 measurements (27.00%)
  19 (19.00%) low mild
  5 (5.00%) high mild
  3 (3.00%) high severe

random_access_1000_100_32
                        time:   [1.3774 ms 1.3786 ms 1.3798 ms]
                        change: [-18.102% -17.807% -17.518%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 6 outliers among 100 measurements (6.00%)
  1 (1.00%) low mild
  1 (1.00%) high mild
  4 (4.00%) high severe

random_access_1000_100_64
                        time:   [2.0736 ms 2.0788 ms 2.0843 ms]
                        change: [-16.603% -16.325% -16.035%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 20 outliers among 100 measurements (20.00%)
  8 (8.00%) high mild
  12 (12.00%) high severe

random_access_1000_1000_2
                        time:   [160.34 µs 161.76 µs 163.36 µs]
                        change: [-23.610% -22.749% -21.871%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 12 outliers among 100 measurements (12.00%)
  4 (4.00%) high mild
  8 (8.00%) high severe

random_access_1000_1000_16
                        time:   [866.60 µs 870.44 µs 874.72 µs]
                        change: [-26.648% -26.278% -25.914%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 5 outliers among 100 measurements (5.00%)
  3 (3.00%) low mild
  2 (2.00%) high mild

random_access_1000_1000_32
                        time:   [1.6365 ms 1.6406 ms 1.6448 ms]
                        change: [-20.851% -20.679% -20.490%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 21 outliers among 100 measurements (21.00%)
  12 (12.00%) low severe
  2 (2.00%) high mild
  7 (7.00%) high severe

random_access_1000_1000_64
                        time:   [3.1399 ms 3.1422 ms 3.1449 ms]
                        change: [-21.191% -20.848% -20.544%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 2 outliers among 100 measurements (2.00%)
  1 (1.00%) high mild
  1 (1.00%) high severe

random_access_1000_1000_128
                        time:   [4.9066 ms 4.9113 ms 4.9162 ms]
                        change: [-16.943% -16.696% -16.472%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 3 outliers among 100 measurements (3.00%)
  3 (3.00%) high mild

random_access_1000_1000_256
                        time:   [8.3928 ms 8.4041 ms 8.4157 ms]
                        change: [-16.351% -16.198% -16.039%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 2 outliers among 100 measurements (2.00%)
  2 (2.00%) high mild

random_access_1000_10000_2
                        time:   [147.21 µs 147.96 µs 148.80 µs]
                        change: [-4.0314% -1.9468% +0.2129%] (p = 0.08 > 0.05)
                        No change in performance detected.
Found 23 outliers among 100 measurements (23.00%)
  2 (2.00%) low severe
  11 (11.00%) low mild
  4 (4.00%) high mild
  6 (6.00%) high severe

random_access_1000_10000_16
                        time:   [777.24 µs 786.42 µs 796.89 µs]
                        change: [-22.667% -21.677% -20.639%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 20 outliers among 100 measurements (20.00%)
  1 (1.00%) high mild
  19 (19.00%) high severe

random_access_1000_10000_32
                        time:   [1.5462 ms 1.5507 ms 1.5548 ms]
                        change: [-21.418% -20.994% -20.560%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 9 outliers among 100 measurements (9.00%)
  3 (3.00%) low severe
  1 (1.00%) low mild
  5 (5.00%) high severe

random_access_1000_10000_64
                        time:   [2.8812 ms 2.8870 ms 2.8929 ms]
                        change: [-27.772% -27.547% -27.326%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 2 outliers among 100 measurements (2.00%)
  2 (2.00%) high mild

random_access_1000_10000_128
                        time:   [6.0885 ms 6.1050 ms 6.1208 ms]
                        change: [-21.714% -21.454% -21.186%] (p = 0.00 < 0.05)
                        Performance has improved.

random_access_1000_10000_256
                        time:   [11.322 ms 11.340 ms 11.358 ms]
                        change: [-20.439% -20.221% -20.019%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 8 outliers among 100 measurements (8.00%)
  6 (6.00%) high mild
  2 (2.00%) high severe

random_access_10000_2_2 time:   [1.5021 ms 1.5041 ms 1.5061 ms]
                        change: [-27.410% -27.063% -26.714%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 8 outliers among 100 measurements (8.00%)
  4 (4.00%) high mild
  4 (4.00%) high severe

random_access_10000_100_2
                        time:   [2.0735 ms 2.0787 ms 2.0838 ms]
                        change: [-18.158% -17.652% -17.171%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 9 outliers among 100 measurements (9.00%)
  7 (7.00%) low mild
  2 (2.00%) high mild

random_access_10000_100_16
                        time:   [9.9218 ms 9.9325 ms 9.9439 ms]
                        change: [-20.404% -20.067% -19.717%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 5 outliers among 100 measurements (5.00%)
  4 (4.00%) high mild
  1 (1.00%) high severe

random_access_10000_100_32
                        time:   [13.933 ms 13.955 ms 13.979 ms]
                        change: [-19.146% -18.788% -18.442%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 1 outliers among 100 measurements (1.00%)
  1 (1.00%) high mild

random_access_10000_100_64
                        time:   [20.636 ms 20.682 ms 20.729 ms]
                        change: [-16.814% -16.548% -16.303%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 2 outliers among 100 measurements (2.00%)
  2 (2.00%) high mild

random_access_10000_1000_2
                        time:   [2.1887 ms 2.1964 ms 2.2039 ms]
                        change: [-19.340% -18.925% -18.496%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 27 outliers among 100 measurements (27.00%)
  23 (23.00%) low severe
  1 (1.00%) high mild
  3 (3.00%) high severe

random_access_10000_1000_16
                        time:   [9.2323 ms 9.2424 ms 9.2539 ms]
                        change: [-31.059% -30.838% -30.635%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 12 outliers among 100 measurements (12.00%)
  4 (4.00%) low mild
  4 (4.00%) high mild
  4 (4.00%) high severe

random_access_10000_1000_32
                        time:   [17.400 ms 17.414 ms 17.430 ms]
                        change: [-20.500% -20.302% -20.118%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 4 outliers among 100 measurements (4.00%)
  1 (1.00%) low mild
  1 (1.00%) high mild
  2 (2.00%) high severe

random_access_10000_1000_64
                        time:   [32.309 ms 32.340 ms 32.375 ms]
                        change: [-20.369% -20.207% -20.061%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 4 outliers among 100 measurements (4.00%)
  3 (3.00%) high mild
  1 (1.00%) high severe

random_access_10000_1000_128
                        time:   [49.962 ms 50.039 ms 50.124 ms]
                        change: [-16.704% -16.469% -16.256%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 3 outliers among 100 measurements (3.00%)
  2 (2.00%) high mild
  1 (1.00%) high severe

random_access_10000_1000_256
                        time:   [83.979 ms 84.111 ms 84.254 ms]
                        change: [-16.684% -16.506% -16.315%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 6 outliers among 100 measurements (6.00%)
  4 (4.00%) high mild
  2 (2.00%) high severe

random_access_10000_10000_2
                        time:   [2.0915 ms 2.1004 ms 2.1101 ms]
                        change: [-17.811% -17.161% -16.488%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 16 outliers among 100 measurements (16.00%)
  2 (2.00%) high mild
  14 (14.00%) high severe

random_access_10000_10000_16
                        time:   [9.8722 ms 9.9084 ms 9.9470 ms]
                        change: [-30.802% -30.487% -30.148%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 7 outliers among 100 measurements (7.00%)
  6 (6.00%) high mild
  1 (1.00%) high severe

random_access_10000_10000_32
                        time:   [17.550 ms 17.577 ms 17.610 ms]
                        change: [-30.931% -30.768% -30.586%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 12 outliers among 100 measurements (12.00%)
  1 (1.00%) low mild
  6 (6.00%) high mild
  5 (5.00%) high severe

random_access_10000_10000_64
                        time:   [31.884 ms 31.984 ms 32.100 ms]
                        change: [-33.385% -33.065% -32.747%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 7 outliers among 100 measurements (7.00%)
  3 (3.00%) high mild
  4 (4.00%) high severe

random_access_10000_10000_128
                        time:   [65.097 ms 65.181 ms 65.272 ms]
                        change: [-23.459% -23.259% -23.070%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 3 outliers among 100 measurements (3.00%)
  1 (1.00%) high mild
  2 (2.00%) high severe

random_access_10000_10000_256
                        time:   [119.16 ms 119.34 ms 119.55 ms]
                        change: [-21.155% -20.956% -20.754%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 5 outliers among 100 measurements (5.00%)
  3 (3.00%) high mild
  2 (2.00%) high severe
```

</details>

File: src/graph_node.rs [hunk 1]
```diff
@@ -11,5 +11,20 @@ pub struct GraphNode<Id: TransactionId> {
...
     /// Unique edges from this node.
     /// The number of edges is the same as the number of forks.
-    pub edges: AHashSet<Id>,
+    pub edges: Vec<Id>,
+}
+
+impl<Id: TransactionId> GraphNode<Id> {
+    /// Returns true if the edge was added successfully.
+    /// Edges are only added if the edge is unique.
+    /// Because we always add edges as we insert, the edge is unique if the
```
Comment (tao-stones) [inline src/graph_node.rs:20]: Looks like a smart move to use `Vec`, just to confirm that `id`s are sorted in `Vec`? Without knowing the actual use cases, I am curious if this is a possible scenario: initially`vec![A, B, C]`, then `try_add_edge(B)`, which should return `false` in original version.

Comment (apfitzge) [reply src/graph_node.rs:20]: Great question.
This is why I originally used a HashSet, I wanted a fast guarantee that the edge was unique.

I recently had the realization that the usage pattern is as follows:
1. insert_transaction
2. loop over all resource (account) accesses in the transaction, check for outstanding uses
3. For each conflicting outstanding use, we mutate the **blocking** transaction node's edges to add the `id` of the transaction we are inserting.

Since `prio_graph` is single-threaded, if transaction `2` conflicts with `1` on account `A` it will be added to the end of the `1`'s `edges`. If it again conflicts with `1` on account `B`, then `1`'s `edges` vec still has `2` at the end, since that was the most recent conflict.
Basically, because we always add edges for the conflicts in the logical order of our graph we can just check the most recent conflict on any blocking node.

Comment (tao-stones) [reply src/graph_node.rs:20]: ah, single-threaded and logical order guarantees this behavior.

**Review by tao-stones — APPROVED** [review]
lgtm
